### ISSUE TITLE
Auras missing

### ISSUE THREAD

**Issue by zludgy** [AzeriteTeam/AzeriteUI#8]
Hi, first of all, congrats!! your UI looks awsome.
However i got a problem with it.
Some auras seems to be missing.
I play marksmanship hunter and i can't see these auras:

-Serpent sting
-Trick Shots
-Precise shots
-Survival of the Fittest
-Trueshot
-Double Tap
-Hunter's Mark

I didn't tried other classes yet.

Maybe there is a blacklist or something like that but unfortunately i can't use it in that state :/

**Comment by goldpaw** [comment]
Currently the aura system shows long duration (or no duration) auras out of combat, and short ones (and some whitelisted ones) while engaged in combat. 

Are all of those hidden once you start fighting too? Most of them should show up! :o 

We're working on a new and better aura system, though, once that won't directly split between combat/nocombat like that, but instead work by a priority system and always show what there is room to show, starting with the most important. Most important usually always being your own DoTs, HoTs and similar.

**Comment by zludgy** [comment]
Yes, all of the ones that i have mentioned are hidden out and in combat.

**Comment by PatTheMav** [comment]
@goldpaw Thanks for the update on the aura system - right now most "timed" buffs (i.e. food buffs, flasks, etc) are not shown as BfA has so many permanent buffs (PVP bonus, event buff, stance, contracts, Mana tracking, etc) that take up all that space.

Also permanent debuffs (e.g. the strain effects at The Brineworks in Stormsong Valley) are never shown in-combat and are visible out-of-combat only if I targeted myself (because they were at least displayed at the target frame).

**Comment by goldpaw** [comment]
This is the intended behavior. There are 7 aura slots on the player frame, and if we include ALL, many classes will have around 9-10 just standing around in one of the BFA cities. So that's why it's filtered this way.

**Comment by PatTheMav** [comment]
@goldpaw That is understood, my point was that right now especially OOC too many "permanent" buffs take up some of those 7 slots, which is why a priority system will be much appreciated.

**Comment by goldpaw** [comment]
Couldn't agree more! Working on a new (or faster for me to make?) approach to the priority system right now, actually. 

The main difference from the current is that all the current auras are parsed before anything is shown, so we'll be able to decide in advance what should be shown based on what's available, instead of each one following the same strict rules as now. 

An example: 
On the live version, I can't see my own HoTs on myself unless you enter combat. Is this because of the combat/nocombat split or because there already are 7 static auras just THERE when I'm standing around in the ally city in BFA? Who knows. Still, my HoTs aren't there, which is bad. 

But with the system I'm making now, the HoTs would get a higher priority rating, and be shown before all the "useless" ones, ensuring it's there even if I'm chilling at the bank. 

I'll probably make 2 settings here, "Strict" and "Adaptive": 

Adaptive meaning it'll always try to fill as many slots as possible, so after the HoTs, DoTs and whatnot the zone buffs would show up, for all chars, maybe except when in combat because that could get messy. It'll adapt, in other words. 

Strict will follow many of the same rules, but hide a LOT more. This will be the complete opposite of what people have been asking for, but it'll be 100% what me and Blakmane wanted the UI to be. So it'll be the "Goldpaw & Blakmane preferred setting" in a way. But both will be there with the click of a button, and I'm fairly certain a lot of people will prefer the Adaptive setting. 

I have no ETA, and even though I'm tempted to say "somewhere between a few hours and some days", I know for a fact I may not always be able to deliver on those promises, since addon work is pretty much squeezed in between everything else in my life. Meaning all unexpected things takes presedence. And I have a lot of those. :o 

Still, the cache system is already live in the current version, though it currently just mimics the old system. But now it's really just a matter of rewriting the filtering part and adding in the menu options. So it's definitely coming! :)

**Comment by goldpaw** [comment]
Renamed and reopened this to reflect the new upcoming aura filter, so others stopping by can see this thread instead of making multiple new! :)

**Comment by PatTheMav** [comment]
I do like the "idea" behind the current system and in combat it works more or less as intended, but there are of course fringe cases (because Blizzard had found no other way to shoehorn those things in - including but not limited to):

* World events and event-based buffs have (at least from the buff system's POV) "unlimited" run time, this includes weekly-events, war mode buffs, "enlisted" buffs
* Stances are "buffs" (e.g. shadowform)
* Some tracking buffs (e.g. mana stones) are applied as buffs with unlimited run time
* Contracts (reputation bonuses) are long-running buffs (7 days)
* Toys and toy-effects
* Lore-effects (e.g. the strain effects/permanent debuffs in Stormsong, "The Gift" by N'Zoth)

I haven't played round with the LUA API in some time, so I don't know if those buffs can be recognized easily (e.g. by runtime) or if you will need to ferry around a big LUA file with all these buff IDs (which is maintenance-hell).

The challenge here is that some if not most of this information has some importance to players and if the priority-based system works as intended, most of these would never be shown.

_Maybe_, just _maybe_ throwing those permanent buffs somewhere else (and keeping the buff bars above the unit frames clean and in spirit of the UI design) is the best move forward. Maybe a little panel that opens with the character window _or_ a mouse-over effect similar to the action bar handling for additional buttons.

I have no perfect solution in mind either, so just throwing around some thoughts and ideas here.

**Comment by goldpaw** [comment]
There are no perfect solutions, but we can get pretty close. Of course the 7 slot limitation on AzeriteUI is a really limiting factor here, but I'm still sure we can get something that works for everybody in the end. 

The various auras with no duration can be tracked by their auraID. We have fairly extensive lists to recognize these things already, and can easily add more now. 

Back in an earlier UI we used to have a category for inventory items we called "Gizmos". This included toys, engineering stuff and so on. Maybe do something similar for auras at some point? Seeing toy buffs is one of the reason we target people in cities in the first place. 

Anyway, the idea of the new system I'm making for AzeriteUI (and all the other upcoming UIs) is that all the auras are known BEFORE we start filtering. So we can make it a lot "smarter" than it is now. For example, only show zone buffs if no toys/HoTs/DoTs are currently active? And so on. Show things we want to see, but always show what we want the most first.

**Comment by goldpaw** [comment]
Should probably couple this with a clickable button that can open a window showing ALL auras, tbh. Sort of like the old consolidation system. Just a way to quickly check if something is there or not. And for the target frame too. That'd sort all the toy/mount buffs out in a not too intrusive way. 

Costs an aura slot, though. Unless I go for just a small, white plus right before the aura row starts. Sort of like the plus we use next to the minimap when MBB (MinimapButtonBag) is loaded, only smaller. That'd let us keep the aura slots, and get fast access to an overview of everything. Combined with the new improved priority filters, we're starting to get something really usable. (?)

**Comment by PatTheMav** [comment]
Make the button appear on hover only maybe, so it can appear outside of the list of seven? 

Also one brain fart I had was "make the aura list scrollable with mouse wheel" but that'd be bad UI as it's not intuitive.

I'll be curious to see how you'll get this filtering solved in efficient LUA code. Death to nested loops! 😁

**Comment by goldpaw** [comment]
Outside was the plan. Sort of like to the left, right before the aura row starts. Not sure about the hover thing, though. Maybe instead make it appear only when more auras than are shown are available? Show it when we need it? That's been the main concept behind everything else in the UI so far. ;) 

When it comes to the lua solving, I'm thinking about building some flags into the cache when parsing the aura list. Something like auraCache.hasHoTs, auraCache.hasZone or something like that, so the filter function deciding priorities and what to show can know about certain other things existing without needing to nest another loop looking through all the other auras once more. I'm generally always opting for reusing tables, and not do any more function calls than I have to. Those two usually boost the performance more than anything else.

**Comment by PatTheMav** [comment]
Hey @goldpaw - to avoid opening another issue, I've noticed that in combat several crucial auras/debuffs aren't displayed on the player unit frame:

* Azshara raid encounter: The wards give a debuff when the player is drained from empowering them - this debuff is not visible
* Several azerite effects from essences as well as traits are never shown in combat

I guess this is all due to the duration filtering and some of those auras being either permanent or having stacks but no true duration.

**Comment by PatTheMav** [comment]
It should be noted at this point that @goldpaw has indeed worked on this issue (and more) in the classic variant and he's reintegrating/adapting those changes to the retail version in a different place around here.. 👀
